# Exercise 2.A Fundamentals of pandas

In [1]:
import numpy as np
import pandas as pd

## Series

In [2]:
aa = [1, 2, 3, 4]
ab = (5, 6, 7, 8)
ac = {9, 10, 11, 12}
ad = {'first': 13, 'second': 14, 'third': 15, 'fourth': 16}

In [3]:
ser1 = pd.Series(aa)
ser2 = pd.Series(ab)
ser3 = pd.Series(list(ac))  # included list() since the order may change any time; TypeError: 'set' type is unordered, a set is always unordered when series requires a fixed order
ser4 = pd.Series(ad)

print(ser1)
print(ser2)
print(ser3)
print(ser4)

0    1
1    2
2    3
3    4
dtype: int64
0    5
1    6
2    7
3    8
dtype: int64
0     9
1    10
2    11
3    12
dtype: int64
first     13
second    14
third     15
fourth    16
dtype: int64


In [4]:
# ser5 = pd.Series(17, 18, 19, 20)  #TypeError: Index(...) must be called with a collection of some kind, 18 was passed; the argument needs to be a collection, like a list, tuple, or dictionary, before being passed through

## DataFrames from Series

In [5]:
df1 = pd.DataFrame(ser1)
df1

,0
0,1
1,2
2,3
3,4


In [6]:
df1.insert(1, "ser2", ser2)
df1.insert(2, "ser3", ser3)
df1

,0,ser2,ser3
0,1,5,9
1,2,6,10
2,3,7,11
3,4,8,12


In [7]:
df1.insert(3, "ser4", ser4)
df1  # when ser4 was added, it automatically assumed that ser4 was going to be numbers too. Since this is not the case, it shows up as NaN; pandas matches values by indexes, which means that it couldn't find the value (0 != first)

,0,ser2,ser3,ser4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN


In [8]:
df1.loc[4] = [5, 9, 12, 0]
df1

,0,ser2,ser3,ser4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN
4,5,9,12,0.0


## Reading Data into a DataFrame

In [11]:
df2 = pd.read_csv("sakila_example.csv")
df2

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE
...,...,...,...,...,...
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON


In [12]:
df2.head(5)

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE


In [13]:
df2.tail(10)

,title,description,category,language,actor_name
252,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,MILLA KEITEL
253,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,GROUCHO DUNST
254,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,BURT TEMPLE
255,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,UMA WOOD
256,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,RIP WINSLET
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON
261,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,JAMES PITT


In [14]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 262 entries, 0 to 261
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   title        262 non-null    object
 1   description  262 non-null    object
 2   category     262 non-null    object
 3   language     262 non-null    object
 4   actor_name   262 non-null    object
dtypes: object(5)
memory usage: 10.4+ KB


There are 262 entries, likely multi-media content. Likely only content starting with 'A' only if sorted by title. It could also be English only content based on what's provided. This could be like IMDB where actors are listed under films.

In [15]:
df2.describe()

,title,description,category,language,actor_name
count,262,262,262,262,262
unique,46,46,15,1,145
top,ARABIA DOGMA,A Touching Epistle of a Madman And a Mad Cow w...,Horror,English,OPRAH KILMER
freq,12,12,55,262,8


Starting from the top, the count for title, description, category, language, and actor_name isn't useful itself other than stating that it has 262 entries.
Unique is useful, meaning across 46 titles, there were 145 different actors spanning across 15 categories in English.
Most actors in Arabia Dogma, assuming that's the description it followed under too, about 12 max. Horror was the most popular category, appearing 55 times. Oprah Kilmer appeared in most medias, across 8 films.

In [17]:
df3 = pd.read_csv("northwind_example.csv")
df3

,OrderID,ProductName,UnitPrice,Quantity
0,10248,Queso Cabrales,14.0,12
1,10248,Singaporean Hokkien Fried Mee,9.8,10
2,10248,Mozzarella di Giovanni,34.8,5
3,10249,Manjimup Dried Apples,42.4,40
4,10249,Tofu,18.6,9
...,...,...,...,...
495,10435,Mozzarella di Giovanni,27.8,10
496,10436,Gnocchi di nonna Alice,30.4,40
497,10436,Wimmers gute Semmelkndel,26.6,30
498,10436,Rhnbru Klosterbier,6.2,24


In [18]:
df3.describe()

,OrderID,UnitPrice,Quantity
count,500.000000,500.000000,500.000000
mean,10341.364000,23.371400,24.330000
std,54.442663,27.334436,18.069702
min,10248.000000,2.000000,1.000000
25%,10294.000000,10.400000,10.000000
50%,10341.000000,15.800000,20.000000
75%,10389.000000,27.800000,32.000000
max,10436.000000,210.800000,120.000000


Since there is OrderID, we can assume that this is either a sales transaction log or product inventory. There are 500 logs, average price per item is $23.37, about 24.33 products ordered per product. The store is very expensive. We can see the quartile breakdown of all products, accurately centered around the median rather than the mean due to some outliers (min vs max). It contains a box and whiskers type analysis rather than a simple summary.

## Coding a DataFrame

In [19]:
test_data = {
    'ID': [1, 2, 3, 4, 5, 6, 7, 8],
    'Name': ['Alicia', 'Bob', 'Charlie', 'Dominic', 'Eve', 'Frank', 'Grace', 'Heidi'],
    'Age': [25, np.nan, 22, 19, 31, 35, np.nan, 18],
    'City': ['NY', 'LA', 'Chicago', 'Houston', 'Phoenix', 'Boston', 'Austin', 'San Diego'],
    'Score': [85, 92, -1, 65, 78, np.nan, 55, 90]
}

In [20]:
df4 = pd.DataFrame(test_data)
df4

,ID,Name,Age,City,Score
0,1,Alicia,25.0,NY,85.0
1,2,Bob,NaN,LA,92.0
2,3,Charlie,22.0,Chicago,-1.0
3,4,Dominic,19.0,Houston,65.0
4,5,Eve,31.0,Phoenix,78.0
5,6,Frank,35.0,Boston,NaN
6,7,Grace,NaN,Austin,55.0
7,8,Heidi,18.0,San Diego,90.0


In [21]:
df4.to_csv("test_data.csv")